raiden_libs/utils/signing.py

```
@@ -56,63 +61,79 @@ def format_int(value, size):
     if isinstance(public_key, PublicKey):
         public_key = public_key.format(compressed=False)
     assert isinstance(public_key, bytes)
-    return encode_hex(keccak(public_key[1:])[-20:])
+    return to_checksum_address(sha3(public_key[1:])[-20:])
 
 
-def private_key_to_address(private_key: str) -> Address:
+def private_key_to_address(private_key: Union[str, bytes]) -> Address:
     """ Converts a private key to an Ethereum address. """
-    return to_checksum_address(
-        public_key_to_address(PrivateKey.from_hex(remove_0x_prefix(private_key)).public_key),
-    )
+    if isinstance(private_key, str):
+        private_key = to_bytes(hexstr=private_key)
+    pk = PrivateKey(private_key)
+    return public_key_to_address(pk.public_key)
 
 
-def address_from_signature(sig: bytes, msg: bytes) -> Address:
+def address_from_signature(data: bytes, signature: bytes, hasher=sha3) -> Address:
     """Convert an EC signature into an ethereum address"""
-    assert len(sig) == 65
+    if not isinstance(signature, bytes) or len(signature) != 65:
+        raise ValueError('Invalid signature, must be 65 bytes')
     # Support Ethereum's EC v value of 27 and EIP 155 values of > 35.
-    if sig[-1] >= 35:
-        network_id = (sig[-1] - 35) // 2
-        sig = sig[:-1] + bytes([sig[-1] - 35 - 2 * network_id])
-    elif sig[-1] >= 27:
-        sig = sig[:-1] + bytes([sig[-1] - 27])
+    if signature[-1] >= 35:
+        network_id = (signature[-1] - 35) // 2
+        signature = signature[:-1] + bytes([signature[-1] - 35 - 2 * network_id])
+    elif signature[-1] >= 27:
+        signature = signature[:-1] + bytes([signature[-1] - 27])
+
+    try:
+        signer_pubkey = PublicKey.from_signature_and_message(signature, data, hasher=hasher)
+        return public_key_to_address(signer_pubkey)
+    except Exception as e:  # pylint: disable=broad-except
+        # coincurve raises bare exception on verify error
+        raise ValueError('Invalid signature') from e
```

Can you raise a specific exception here instead of `ValueError`?